### ISSUE TITLE
[Bug] `linear` operator produces incorrect results on Ascend NPU (all-`!` garbage output in vLLM MoE inference)

### ISSUE THREAD

**Issue by Lxiparer** [FlagOpen/FlagGems#5657]
## Summary

The `linear` operator (`flag_gems.ops.linear.linear`, registered under the key `linear`) produces **numerically wrong results on Ascend NPU (ascend910_9391 / 910B-class)**. When FlagGems is enabled under vLLM on Ascend, every projection layer of a Qwen3.5-MoE model routes through this kernel, and the corruption propagates to the LM head, so **the model emits nothing but exclamation marks** (`!!!!!!!!!!!!!!!!`) for every request.

The bug is isolated: with a fixed prompt and `temperature=0`, blacklisting **only** `linear` (leaving every other FlagGems op active) restores correct output. Re-enabling `linear` reproduces the garbage output deterministically.

`aten::mm` / `aten::addmm` were already disabled in our deployment, but `linear` is a **separately registered op** and was not covered, which is why the corruption persisted until we traced it down.

## Environment

| Component | Version |
|---|---|
| FlagGems | `5.3.4.post1.dev128+g05cd332b3` (commit `05cd332b373b7f87f96962fe1c48c482dc654b0e`) |
| Backend | Ascend NPU, `SOC_VERSION=ascend910_9391` |
| torch | `2.10.0` |
| torch_npu | `2.10.0` |
| triton | provided by **flagtree** `0.6.0+ascend.git317f15a4` (imports as `triton`, reports `__version__ == 3.5.1`); `triton_ascend 3.2.1` also installed but shadowed |
| CANN | `9.0.0` |
| Framework | vLLM `0.20.2`, vllm-ascend `0.20.2rc1`, vllm-plugin-FL `0.2.0` |
| Model | Qwen3.5-MoE (`Qwen3_5MoeForConditionalGeneration`, `qwen3_5_moe`), TP=2, `--enforce-eager` |

Offending op introduced by: **PR #3797 — `[KernelGen][Nvidia] Add linear operator with Triton kernel`** (commit `41ace9c3bef5702a03fee9e6c55b18541f2b8e7e`). As the tag indicates, the kernel was generated and tuned for **Nvidia** and has no Ascend-specific validation.

## Root-cause analysis

Two concrete problems in `src/flag_gems/ops/linear.py`:

### 1. No Ascend tuning configuration exists for `linear`

`linear_kernel` is decorated with:

```python
@libtuner(
    configs=runtime.get_tuned_config("linear"),
    key=["M", "N", "K"],
    strategy=["align32", "align32", "align32"],
    ...
)
```

`linear` tune configs are present for `_nvidia`, `_enflame`, `_iluvatar`, `_mthreads`, `_tsingmicro`, but **`src/flag_gems/runtime/backend/_ascend/tune_configs.yaml` has no `linear:` entry**. On Ascend the kernel therefore runs with block sizes / tuning parameters that were never validated for this hardware.

### 2. The Triton GEMM is not Ascend-safe

The inner loop uses masked block loads plus `tl.dot(..., allow_tf32=False)`:

```python
for k in range(0, tl.cdiv(K, BLOCK_SIZE_K)):
    input_mask_k  = offs_k < (K - k * BLOCK_SIZE_K)
    weight_mask_k = offs_k < (K - k * BLOCK_SIZE_K)
    a = tl.load(input_ptrs,  mask=input_mask,  other=0.0)
    b = tl.load(weight_ptrs, mask=weight_mask, other=0.0)
    accumulator += tl.dot(a, b, allow_tf32=False)
    input_ptrs  += BLOCK_SIZE_K * stride_ik
    weight_ptrs += BLOCK_SIZE_K * stride_wk
```

On this Ascend setup the kernel is compiled by the **flagtree** Ascend fork (`0.6.0+ascend.git317f15a4`, importing as `triton`), not by upstream Triton or `triton_ascend`. This hand-written tiled GEMM produces wrong values (as opposed to crashing): the result is silent numerical corruption, not an exception — the model runs to completion and returns tokens, but they are garbage. This class of "generated for Nvidia, wrong on Ascend" failure is the most dangerous kind because it passes startup and only surfaces as bad output quality. Because the compile stack is flagtree, the root cause may lie either in the kernel (untuned block sizes / masking) or in how flagtree lowers this `tl.dot`-based GEMM for Ascend; both should be checked when validating a fix.

### In-context (vLLM)

```bash
VLLM_PLUGINS=fl VLLM_FL_PLATFORM=ascend \
vllm serve /path/to/Qwen3.5-MoE \
  --tensor-parallel-size 2 --enforce-eager --max-model-len 4096 \
  --served-model-name qwen --port 7023
```

```bash
curl -s http://127.0.0.1:7023/v1/completions \
  -H 'Content-Type: application/json' \
  -d '{"model":"qwen","prompt":"请用一句话介绍北京。","max_tokens":40,"temperature":0}'
```

**With `linear` enabled** → output is `!!!!!!!!!!!!!!!!` (all exclamation marks).
**With `linear` blacklisted** (everything else identical) → output is correct:
`北京是中国的首都，也是一座融合了深厚历史底蕴与现代都市活力的国际化大都市。`

### Minimal standalone check (recommended for a unit test)

```python
import torch, torch_npu
torch.npu.set_device(0)
import flag_gems
flag_gems.enable(once=True)

torch.manual_seed(0)
x = torch.randn(512, 2048, device="npu", dtype=torch.float16)
w = torch.randn(4096, 2048, device="npu", dtype=torch.float16)
b = torch.randn(4096, device="npu", dtype=torch.float16)

ref  = torch.nn.functional.linear(x, w, b)          # aten reference
gems = flag_gems.ops.linear.linear(x, w, b)         # FlagGems kernel

torch.testing.assert_close(gems, ref, rtol=1e-2, atol=1e-2)
# Expect: large mismatch on Ascend (correct on Nvidia)
```

## Bisection evidence

We started by blacklisting the full set of FlagGems ops that were actually dispatched during a run (`linear, silu, sigmoid, softplus, cumsum, rsqrt, cos, sin, sum, clamp, pad, ...`) → output correct. We then narrowed the blacklist down to a **single op, `linear`**, restoring every other op → output still correct. Removing `linear` from the blacklist → garbage returns. This uniquely identifies `linear` as the cause.

## Impact

- **Severity: high.** Any model on Ascend that routes `aten::linear` through FlagGems (most transformer inference) gets silently corrupted outputs, with no crash and no warning.
- Easy to miss: startup succeeds, HTTP 200 is returned, only the generated text is wrong.

## Suggested fixes (in order of preference)

1. **Add and validate an Ascend tune config** for `linear` in `src/flag_gems/runtime/backend/_ascend/tune_configs.yaml`, and add an Ascend accuracy test (see standalone snippet). Only enable the Triton kernel on Ascend once it passes `assert_close` against `F.linear`.
2. **Until validated, do not register `linear` as an Ascend override** — fall back to the native `aten::linear` / `torch_npu` path on Ascend. A per-backend registration guard is cleaner than requiring every downstream integrator to blacklist it.
3. Review other `[KernelGen][Nvidia]`-generated ops added after 5.0.2 for the same "no Ascend tune config + unverified accuracy" pattern; they are latent silent-corruption risks on Ascend.

## Current workaround

Disable the op on Ascend via the integrator's blacklist (in our case vllm-plugin-FL's `dispatch/config/ascend.yaml` `flagos_blacklist`):

```yaml
flagos_blacklist:
  - linear
```

This makes `linear` fall back to the native implementation and produces correct output.
